### ISSUE TITLE
Improve Spyder 3 theme

### ISSUE THREAD

**Issue by goanpeca** [spyder-ide/spyder#2490]
We can start listing any improvements that need to be done to the current Spyder 3 theme using QtAwesome icons.

Here my concerns.

The following icons should be filled icons and not contour ones. Most of the interface uses filled icons and we should strive for coherence and unity (besides it looks better)
- New File
- Save
- Save as

Other comments
- Maximize/Unmaximize icon: This icons are not vertically centered
- File explorer. Isnt this supposed to be using iconic fonts as well?

---

We need a coherent color theme for Spyder, so that the object inspector and the spyder colors at least match.

---

I really think that Spyder 3 theme should be the default in 3.0

**Comment by ccordoba12** [comment]
My main concerns (for now) are:

1. *Run cell* and *Run cell and advance*. I gave @SylvainCorlay some suggestions on how to make them simpler.
2. IPython console: I really think we should start using the Jupyter icon instead of the current one.
3. Python Console: The *Spyder 3* icon is misleading because it's using the one commonly associated to operating system terminals. I think we should continue using the *Spyder 2* theme icon, i.e. the blue and yellow Python icon.
4. Python Path manager: We should find a better icon for it, specially if we follow my suggestion 3 (because *Spyder 3* is using the blue and yellow icon for it).

----

About @goanpeca's concerns:

> We need a coherent color theme for Spyder, so that the object inspector and the spyder colors at least match.

This is not related to the new icons, so please open a new issue for it. However, I don't know how to use the same theme in the object inspector than the one in the Editor. We would probably need to generate multiple css files for it, using something like `less`. @SylvainCorlay knows more about it.

> I really think that Spyder 3 theme should be the default in 3.0

Not until the theme is finished :-) But yes, it'll probably be the default.

**Comment by SylvainCorlay** [comment]
- Regarding the run-cell icons, I am totally +1, as well as the font bearing adjustments for the alignment issues.

- Regarding the filled versions of the font-awesome icons, I am not sure it is my place to provide them. Please vote up on the font-awesome issues. https://github.com/FortAwesome/Font-Awesome/issues/1405

- For the file explorer, tab-minimize/maximize icons, these are native icons, and I am working on overloading the right things in QStyle, but it is more on the long-run. Before creating a native Qt theme based on font-awesome, I will move around icon files and rename things to free-desktop compatible names.

-Another thing that I want to do, is stop using the icon manager in the outline explorer, and only use qtawesome with coloring based on the syntax coloring theme.

**Comment by SylvainCorlay** [comment]
Ok, I addressed the alignment issues and the run-cells.

**Comment by goanpeca** [comment]
Ok so... I really dislike the current cell icons... they do not fit in the theme (they are 3D) and they have too much stuff going on... that is not the point of the font awesome style, which should be more like:
- Clean
- Flat
- Simple

Also I have stressed this point many times before and I will do it again, we need to be consistent in the use of filled icons (vs contour icons). `New File`, `Save`, and `Save all` icons are contour icons and should be filled. Since font-awesome does not provide a filled floppy disk, we need to ship this as part of the spyder icon theme.

Changes needed to be consistent:
- New file should use [this](http://fortawesome.github.io/Font-Awesome/icon/file/)
- Save and save all: Make a new icon based on the ideas [here](https://github.com/FortAwesome/Font-Awesome/issues/1405), (@spyder-ide/core-developers vote +1 so in the future we can use a builtin icon)
- Cell icons: We need to remove borders, so I would revert to a white background with a cell rectangle (no borders) and maybe we can change the minuscule red arrow for a more evident [normal arrow](http://fortawesome.github.io/Font-Awesome/icon/arrow-down/) going in the center to the bottom 

@SylvainCorlay can you make this happen :-) ?

**Comment by ccordoba12** [comment]
My tasks, from issue #2527:

- [ ] Add a menu entry to our main contextual menu so that people can decide the size of icons in toolbars. Suggested options: small: 16x16, medium: 24x24, large: 32x32, big: 48x48.
- [ ] Add the possibility to turn off hinting in qtawesome for Linux.

**Comment by ccordoba12** [comment]
@SylvainCorlay, I agree with @goanpeca in his three proposed tasks. Tasks 1 and 2 should be an easy fix, but task 3 is still the more important one, i.e. those bloody cell icons! :-p The border on those icons looks really ugly on Windows and Mac, so we need to remove it. But without border, then a white background seems like out of place too. Let's make them simpler :-) For that I'm going to upload some sketches I have in mind, so we can have something to discuss with pictures and not just words :-)

About my tasks: I can create a PR to fix task 1. but you've got help me with task 2 because I don't know how to do it :-)

**Comment by ccordoba12** [comment]
Another task:
- [ ] Turn on Hdpi on Mac (see PR #2718).

**Comment by ccordoba12** [comment]
And another one (sorry, before I forget it):
- [ ] Change IPython icon to the Jupyter one

----

I know we have discussed this before in person with @SylvainCorlay, but the current IP icon doesn't render well. So changing it to the Jupyter one (which seems closer to the FontAwesome aesthetics) seems the best option to me. Besides, we're using the Jupyter qtconsole package, so it's not bad to reflect that on the UI too.

And by Jupyter icon, I mean something really simple, like the favicon present in https://try.jupyter.org/

**Comment by SylvainCorlay** [comment]
It seems that PyQt5 might be required to have the HdPi on mac. 

With PyQt4, icons are on my retina display in the toolbars and push buttons are pixelated, while they are perfectly crisp elsewhere (like in the preference menu).

Starting with Qt5, we should be able to set `app.setAttribute(QtCore.Qt.AA_UseHighDpiPixmaps)` at the level of the QApplication.
Besides, it some threads mention that the python.app's Info.plist might have to be edited to have
```
 <key>NSPrincipalClass</key>
<string>NSApplication</string>
<key>NSHighResolutionCapable</key>
<true/>
```
to work from the bootstrap script.

**Comment by ccordoba12** [comment]
@Nodd said this on issue #2908:

I activated the spyder3 theme for icons. The unification of all the icons is really great, but some lack of colors make it harder to grasp the information conveyed by the icon rapidely. Examples:

* The icon indicating an error in the line in the editor margin should be red, not black
* The icons for both private and public methods in the Outline are both round, white on green. It's harder to distinguish between the to with a quick glance. Also the icon for private methods looks like a green no entry sign, that's disturbing. The hyphen could be at least a bit to the bottom so that it looks more like an underscore.
* The stop icon on consoles could be redish
* The parameter icons could be dark grey instead of full black
* TODO checkmarks in the margins have the same orange as the warning icon. The blue of the old icons was good.
* Quit icon should be red
* A propos should be blue (the `i`)
* Documentation should be red (the buoy)
* the rotation arrow for reboot is usually blue

The main point is that color conveys lots information in addition to the shape. Some color are more or less standard, and reusing these standard colors help finding the correct entry quickly. Too many icons are plain black by default: each one look great, but as a whole there's just black everywhere. Nothing looks more like a round icon than another round icon :)

**Comment by ccordoba12** [comment]
@SylvainCorlay, this is my proposal for simpler cell icons:

![cell](https://cloud.githubusercontent.com/assets/365293/12211346/46533b28-b62e-11e5-82d7-13d5fc7cb046.png)

So the idea is to use a rectangle with transparent background (ilke the *New File* icon, i.e. not white background), and use lines with the same width of the borders (or maybe a little bit thinner but not that much) to signal the presence of a cell.

*Note*: My icon misses the red arrow for *run and advance* and the vertical bar next to the play button. But I agree with them. So @SylvainCorlay, please add them back after the redesing :-)

**Comment by ccordoba12** [comment]
Another task

- [ ] Remove the plus sign from the *New file* icon. I don't like it, I think it's fine without that sign.

**Comment by SylvainCorlay** [comment]
Here is the current version of the run toolbar. The run cell and run cell and advance are essentially the same as what you are proposing.

<img width="428" alt="toolbar" src="https://cloud.githubusercontent.com/assets/2397974/12212138/db97bc1c-b634-11e5-8cad-94995584e424.png">

A change I wanted to make to the run-cell and advance is change the red arrow to something thicker. 

The main critic of Gonzalo on this specific icon was that it is not a flat design.

**Comment by ccordoba12** [comment]
The main thing I'm proposing is to remove the white background of the cell icons, and leave it transparent (like the *New file* icon).

I'm also suggesting to make the cell borders of the same width of those the represent the file.

**Comment by SylvainCorlay** [comment]
> I'm also suggesting to make the cell borders of the same width of those that represent the file.

Same color too?

**Comment by ccordoba12** [comment]
Yep, let's make those icons as flat and simple as possible.

**Comment by SylvainCorlay** [comment]
Removing the white background and using the same color for the cell border (but not yet the same width) gives:

<img width="427" alt="toolbar2" src="https://cloud.githubusercontent.com/assets/2397974/12212818/8a94e5a8-b63c-11e5-97c6-ab2e9b5304f3.png">

Do you already find it better?

**Comment by ccordoba12** [comment]
> Do you already like it better?

Yes, I do! But I have another suggestion (that you probably won't like): What if we remove the vertical (green) bar next to the play button? I know you were the one who proposed its addition, and I understand what you try to communicate with it, but:

1. That bar makes the cell icons harder to grasp visually because:
  * It is placed so close to the border that you have to make a bit of an effort to distinguish bar from border.
  * The bar presence doesn't help you to see that there's a play button behind it. Again, you have to do an extra effort to see that button.
2. We could move the play button a bit more to the right without that bar, which (I think) would make the cell icons look nicer.

> I also want to use the roughly same arrow as in step-out for the run-cell and advance

Well thought! I like that arrow too.

----

But please don't focus only on our cell icons, and try to address the rest of the comments on this issue :-)

**Comment by SylvainCorlay** [comment]
> Yes, I do! But I have another suggestion (that you probably won't like): What if we remove the vertical (green) bar next to the play button? I know you were the one who proposed its addition, and I understand what you try to communicate with it, but:

> 1. That bar makes the cell icons harder to grasp visually because:
>   * It is placed so close to the border that you have to make a bit of an effort to  distinguish bar from border.
>  * The bar presence doesn't help you to see that there's a play button behind it.  Again, you have to do an extra effort to see that button.
> 2. We could move the play button a bit more to the right without that bar, which (I think) would make the cell icons look nicer.

I don't agree that we should remove the vertical bar. The notebook has that vertical bar for run-cell-and-advance. A regular play button would be confusing because we have too many of them in the toolbar, and it could give the wrong impression that it is some other run-configurations or that it would play the whole document.
I can make is more easily distinguished from the page border.

**Comment by ccordoba12** [comment]
Ok then, please try to make the buttons more visually discernible :-)

**Comment by zeehio** [comment]
Have you thought of making toolbars a bit thinner? When I am on my laptop I feel that a lot of space could be used for reading code (editor, console) and it is spent on toolbar icons ~~and padding~~ instead. 

*Update: By changing the "Qt Windows Style" from GTK+ to Cleanlooks I can read a couple more lines of code.*

Anyway, I did a quick test changing [this line](https://github.com/spyder-ide/spyder/blob/5f4613d5f05e679b53e0a43468d3afa8fe14901f/spyderlib/app/spyder.py#L502) from 24 to 16. I could not change the width of the toolbar but the icons still look pretty readable to me.

`iconsize=24`:

![icon_24](https://cloud.githubusercontent.com/assets/75441/12963726/8e46b274-d04c-11e5-82f5-72106768c53c.png)

`iconsize=16`: (toolbar width did not change, I did not look further on how to change it)
![icon_16](https://cloud.githubusercontent.com/assets/75441/12963722/8995dade-d04c-11e5-81b6-a76e81d1f3af.png)

(Spyder beta installed through conda/spyder-ide channel, on a Linux-64 machine)

(Thanks for your time and effort. Spyder is great as it is and I believe you are all doing a great work, this is just some safe-to-ignore constructive feedback :sweat_smile:)

**Comment by SylvainCorlay** [comment]
Thanks @zeehio for your feedback.

I very much agree with you since qtawesome icons are pixel perfect at sizes that are multiple of 16 pixels on low dpi screens and sizes that are multiple of 8 pixels for high dpi screens.

However, we kept the original 24 pixel icon size for the toolbar because of the old icon theme that looks better with this setting. We could make this parameter dependent of the theme.

**Comment by EvilMav** [comment]
Unfrtunately, the (sexy =) ) new icons do not work well for dark desktop themes - like Breeze Dark and Maja Dark in Plasma5, where they are barely readable. The old ones did. Any chance an inverted color version could be added at some point?

**Comment by ccordoba12** [comment]
This is matter of changing a parameter in QtAwesome. We'll work on it for Spyder 3.2

**Comment by j9ac9k** [comment]
Came across this thread while trying to figure out a good way to have colors of qtawesome icons work well for both dark themes (breeze-dark) and standard themes, @ccordoba12  you mentioned changing a parameter in QtAwesome to to deal with this, could you elaborate or point me in the right direction with how to go by addressing this?

**Comment by goanpeca** [comment]
@j9ac9k moving forward, spyder will no longer use QtAwesome, we will be implementing a new theme and provide the ability to users to change this theme as they see fit. Icons will be based on SVGs colored by the global color theme that the user has chosen.

**Comment by j9ac9k** [comment]
@goanpeca thanks for the response, if you don't mind me asking, what was the decision behind not using QtAwesome (if discussed in another place, feel free to just send me the link).

**Comment by goanpeca** [comment]
Tha main issue is Windows 10 does not allow to load fonts from folders outside what Microsoft has defined, and spyder was having a lot of issues. Creating fonts for custom icons to use as iconic fonts with QtAwesome was a major pain point, so we are much better of with svg files organized in a theme folder.

**Comment by ccordoba12** [comment]
This is going to be fixed once we add a dark theme for Spyder, so I'm closing this issue as a duplicate of issue #2350.
